test/metabase/models/setting_test.clj

```diff
@@ -142,14 +155,54 @@
     (is (= true
            (test-setting-calculated-getter)))
     (is (= true
-           (setting/user-facing-value :test-setting-calculated-getter)))))
+           (setting/user-facing-value :test-setting-calculated-getter))))
+
+  (testing "`user-facing-value` will initialize pending values"
+    (is (some? (setting/user-facing-value custom-init-setting)))))
 
 (deftest do-not-define-setter-function-for-setter-none-test
   (testing "Settings with `:setter` `:none` should not have a setter function defined"
     (testing "Sanity check: getter should be defined"
       (is (some? (resolve `test-setting-calculated-getter))))
     (is (not (resolve `test-setting-calculated-getter!)))))
 
+(defn- get-setting-via-db []
+  (#'setting/read-setting :test-setting-custom-init))
+
+;; TODO: I suspect we're seeing stale values in CI due to parallel tests or state persisting between runs
+;;  We should at least make this an error when running locally without parallelism.
+(defn- clear-setting-if-leak! []
+  (when-let [existing-value (some? (get-setting-via-db))]
+    (log/warn "Test environment corrupted, perhaps due to parallel tests or state kept between runs" existing-value)
+    (setting/set! :test-setting-custom-init nil :bypass-read-only? true)))
+
+(deftest setting-initialization-test
+  (testing "The value will be initialized and saved"
+    (clear-setting-if-leak!)
+    (let [val (setting/get custom-init-setting)]
+      (is (some? val))
+      (is (= val (test-setting-custom-init)))
+      (is (= val (get-setting-via-db))))))
+
+(deftest validate-without-initialization-test
+  (testing "Validation does not initialize the setting"
+    (clear-setting-if-leak!)
+    (setting/validate-settings-formatting!)
+    (is (= nil (get-setting-via-db)))))
+
+(deftest init-requires-db-test
+  (testing "We will fail instead of implicitly initializing a setting if the db is not ready"
```

Comment:
To keep this test isolated, you could set `test-setting-custom-init`'s value to `nil` at the beginning. Other tests should be using `mt/discard-settings-changes` but better to be safe than sorry